lib/source-param-middleware.js

```diff
@@ -1,25 +1,68 @@
 'use strict';
 
 const defaults = require('lodash/defaults');
+const fetchSystemCodes = require('./fetch-system-codes');
 const httpError = require('http-errors');
 
 module.exports = sourceParam;
 
 module.exports.defaults = {
-	errorMessage: 'The source parameter is required and should be a valid system code'
+	cmdbApiKey: null,
+	errorMessage: 'The source parameter is required and must be a valid system code',
+	log: console,
+	pollInterval: 300000, // 5 minutes
+	verifyUsingCmdb: true
 };
 
+// Create a middleware function which requires the
+// request to contain a valid source parameter
 function sourceParam(options) {
 	options = defaults({}, options, module.exports.defaults);
+	let validSystemCodes;
 
-	return (request, response, next) => {
-		if (!isValidSourceParam(request.query.source)) {
+	// If we're verifying with CMDB then start
+	// polling for a list of valid system codes
+	if (options.verifyUsingCmdb) {
+		pollSystemCodes();
+		setInterval(pollSystemCodes, options.pollInterval);
+	}
+
+	// Function used to poll for system codes
+	function pollSystemCodes() {
+		// Note we need to return here to allow tests
+		// to hook into the promise
+		return fetchSystemCodes(options.cmdbApiKey)
+			.then(codes => {
+				validSystemCodes = codes;
+			})
+			.catch(error => {
+				// If the poll errors then we don't throw,
+				// we just start letting any valid source
+				// parameter through
+				validSystemCodes = undefined;
+				options.log.error(`CMDB error: ${error.message}`);
+			});
+	}
+
+	// The generated middleware function
+	function middleware(request, response, next) {
+		// Check for a valid system code format
+		if (!isValidSystemCodeFormat(request.query.source)) {
 			return next(httpError(400, options.errorMessage));
 		}
-		next();
-	};
+		// If we have a copy of the system codes array,
+		// check the source parameter against it
+		if (validSystemCodes && !validSystemCodes.includes(request.query.source)) {
```

Comment:
Should we log out every request where `validSystemCodes` is falsey if `verifyUsingCmdb` is set to true? That way they can see every request which came in whilst the validation was skipped.